### PR TITLE
Add multi source semantic tests

### LINKED ISSUE #9210: [Testing] Add semantic testing framework for testing imports from multiple files
## Abstract

[ ] Add support for parsing multiple files and expectations
[ ] https://github.com/ethereum/solidity/issues/9216 tracks support for tests making use of this feature

### PR DESCRIPTION
**Pull request by bshastry**
Fixes #9210

### CODE REVIEW

File: test/tools/isoltest.cpp [hunk 1]
```diff
@@ -21,17 +21,13 @@
...
 #include <test/InteractiveTests.h>
 #include <test/EVMHost.h>
 
-#include <boost/algorithm/string.hpp>
 #include <boost/algorithm/string/replace.hpp>
 #include <boost/filesystem.hpp>
-#include <boost/program_options.hpp>
 
 #include <cstdlib>
-#include <fstream>
```
Comment (bshastry) [inline test/tools/isoltest.cpp:34]: These are unused includes that clion pointed out

File: test/ExecutionFramework.h [hunk 2]
```diff
@@ -68,6 +68,15 @@ class ExecutionFramework
 		std::map<std::string, Address> const& _libraryAddresses = std::map<std::string, Address>()
 	) = 0;
 
+	virtual bytes const& multiSourceCompileAndRunWithoutCheck(
```
Comment (chriseth) [inline test/ExecutionFramework.h]: Shouldn't this be called from the regular "compileAndRun" function with just a single entry in the `_sourceCodeMap`?

File: test/ExecutionFramework.h [hunk 3]
```diff
@@ -68,6 +68,15 @@ class ExecutionFramework
...
+		std::map<std::string, std::string> const& _sourceCodeMap,
```
Comment (chriseth) [inline test/ExecutionFramework.h]:
```suggestion
		std::map<std::string, std::string> const& _sources,
```
(the type of the variable should not be encoded in the name)

File: test/TestCaseReader.h [hunk 4]
```diff
@@ -23,6 +23,20 @@
...
 {
+
+/**
+ * A map for registering source names that also contains the last source name in a test case.
+ */
+struct SourceMap
+{
+	std::map<std::string, std::string> m_sourceMap;
+	std::string m_lastSourceName;
+	std::string last()
```
Comment (chriseth) [inline test/TestCaseReader.h]: Is that function really needed?

File: test/TestCaseReader.h [hunk 5]
```diff
@@ -23,6 +23,20 @@
 
 namespace solidity::frontend::test
 {
+
+/**
+ * A map for registering source names that also contains the last source name in a test case.
+ */
+struct SourceMap
+{
+	std::map<std::string, std::string> m_sourceMap;
```
Comment (chriseth) [inline test/TestCaseReader.h]: public members in structs do not have a `m_` prefix

File: test/TestCaseReader.h [hunk 6]
```diff
@@ -23,6 +23,20 @@
...
+	std::string m_lastSourceName;
```
Comment (chriseth) [inline test/TestCaseReader.h]: We could abstract the "last" detail away and just say `mainSourceName`?

File: test/libsolidity/SemanticTest.cpp [hunk 7]
```diff
@@ -44,7 +44,11 @@ SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVer
 	EVMVersionRestrictedTestCase(_filename),
 	m_enforceViaYul(enforceViaYul)
 {
-	m_source = m_reader.source();
+	m_source = m_reader.sources().m_sourceMap;
```
Comment (chriseth) [inline test/libsolidity/SemanticTest.cpp]: Maybe better to copy the whole struct?

File: test/libsolidity/SolidityExecutionFramework.cpp [hunk 8]
```diff
@@ -31,8 +31,34 @@ using namespace solidity::frontend;
...
 using namespace std;
 
+bytes SolidityExecutionFramework::multiSourceCompileContract(
+	std::map<std::string, std::string> const& _sourceCode,
+	string const& _contractName,
+	map<string, Address> const& _libraryAddresses
+)
+{
+	map<string, string> sourceCodeMap = _sourceCode;
+	for_each(
```
Comment (chriseth) [inline test/libsolidity/SolidityExecutionFramework.cpp]: Is it not better to use a range-based for loop here?

File: test/TestCaseReader.h (hunk 4 above)
Comment (bshastry) [reply test/TestCaseReader.h]: No, I'll remove it.

File: test/ExecutionFramework.h (hunk 2 above)
Comment (chriseth) [reply test/ExecutionFramework.h]: I meant that there should only be one function to implement in derived classes and this is the multi-source variant. Maybe we can even just change the signature of `compileAndRunWithoutCheck` instead? Execution frameworks / compilers that do not support multi-source inputs can just assert that there is only one file.

File: test/TestCaseReader.h [hunk 9]
```diff
@@ -23,6 +23,16 @@
 
 namespace solidity::frontend::test
 {
+
+/**
+ * A map for registering source names that also contains the last source name in a test case.
```
Comment (chriseth) [inline test/TestCaseReader.h]:
```suggestion
 * A map for registering source names that also contains the main source name in a test case.
```

File: test/CommonSyntaxTest.cpp [hunk 10]
```diff
@@ -60,7 +60,7 @@ CommonSyntaxTest::CommonSyntaxTest(string const& _filename, langutil::EVMVersion
 	EVMVersionRestrictedTestCase(_filename),
 	m_evmVersion(_evmVersion)
 {
-	m_sources = m_reader.sources();
+	m_sources = m_reader.sources().sourceMap;
 	m_expectations = parseExpectations(m_reader.stream());
```
Comment (cameel) [inline test/CommonSyntaxTest.cpp]: I think these could be moved to the initializer list above.

File: test/ExecutionFramework.h [hunk 11]
```diff
@@ -76,7 +85,13 @@ class ExecutionFramework
 		std::map<std::string, Address> const& _libraryAddresses = std::map<std::string, Address>()
 	)
 	{
-		compileAndRunWithoutCheck(_sourceCode, _value, _contractName, _arguments, _libraryAddresses);
+		multiSourceCompileAndRunWithoutCheck(
+			std::map<std::string, std::string>({{ "", _sourceCode }}),
```
Comment (cameel) [inline test/ExecutionFramework.h]:
```suggestion
			std::map<std::string, std::string>{{"", _sourceCode}},
```
I think this should work too:
```suggestion
			{{"", _sourceCode}},
```

File: test/ExecutionFramework.h [hunk 12]
```diff
@@ -68,6 +68,15 @@ class ExecutionFramework
 		std::map<std::string, Address> const& _libraryAddresses = std::map<std::string, Address>()
 	) = 0;
 
+	virtual bytes const& multiSourceCompileAndRunWithoutCheck(
+		std::map<std::string, std::string> const& _sources,
+		u256 const& _value = 0,
+		std::string const& _contractName = "",
+		bytes const& _arguments = bytes(),
+		std::map<std::string, Address> const& _libraryAddresses = std::map<std::string, Address>()
```
Comment (cameel) [inline test/ExecutionFramework.h]:
```suggestion
		bytes const& _arguments = {},
		std::map<std::string, Address> const& _libraryAddresses = {}
```

Same in functions above and below and in `SolidityExecutionFramework.h`.

File: test/TestCaseReader.cpp [hunk 13]
```diff
@@ -39,9 +39,9 @@ TestCaseReader::TestCaseReader(string const& _filename):
 
 string const& TestCaseReader::source()
 {
-	if (m_sources.size() != 1)
+	if (m_sources.sourceMap.size() != 1)
 		BOOST_THROW_EXCEPTION(runtime_error("Expected single source definition, but got multiple sources."));
-	return m_sources.begin()->second;
+	return m_sources.sourceMap.begin()->second;
```
Comment (cameel) [inline test/TestCaseReader.cpp]: This is equivalent, would make it clearer that you're returning the main file and also works as a sanity check that this file is present in the source map:
```suggestion
	return m_sources.sourceMap.at(m_sources.mainSourceFile);
```

File: test/TestCaseReader.cpp [hunk 14]
```diff
@@ -145,8 +146,9 @@ pair<map<string, string>, size_t> TestCaseReader::parseSourcesAndSettingsWithLin
 		else
 			throw runtime_error(string("Expected \"//\" or \"// ---\" to terminate settings and source."));
 	}
+	// Register last source
 	sources[currentSourceName] = currentSource;
-	return { sources, lineNumber };
+	return { { sources, currentSourceName }, lineNumber };
```
Comment (cameel) [inline test/TestCaseReader.cpp]:
```suggestion
	return {{sources, currentSourceName}, lineNumber};
```
I generally only see this style used for function braces in the codebase. The coding style doesn't say anything about it (only not to add spaces inside parentheses) - maybe we should update it to be clear about `[]` and `{}` too?

Also, I think you need an explicit `move` here (compiler won't do it automatically):
```suggestion
	return {{move(sources), move(currentSourceName)}, lineNumber};
```

File: test/TestCaseReader.cpp [hunk 15]
```diff
@@ -118,6 +118,7 @@ pair<map<string, string>, size_t> TestCaseReader::parseSourcesAndSettingsWithLin
 		{
 			if (boost::algorithm::starts_with(line, sourceDelimiterStart) && boost::algorithm::ends_with(line, sourceDelimiterEnd))
 			{
+				// If statement registers previously parsed source
 				if (!(currentSourceName.empty() && currentSource.empty()))
 					sources[currentSourceName] = std::move(currentSource);
```
Comment (cameel) [inline test/TestCaseReader.cpp]: I'd drop this comment. There are so many `if`s here that it actually made it more confusing to me :)

Or, if you really think it's necessary, I'd reword it a bit and move it inside. I'd also move the negation inside parenthesis since this is the part that actually confused me the most (I did not notice the parenthesis and read it as `!A.empty() && B.empty()`).
```suggestion
				if (!currentSourceName.empty() || !currentSource.empty())
					// Finalize the previous file before starting a new one
					sources[currentSourceName] = std::move(currentSource);
```

File: test/TestCaseReader.cpp [hunk 16]
```diff
@@ -145,8 +146,9 @@ pair<map<string, string>, size_t> TestCaseReader::parseSourcesAndSettingsWithLin
 		else
 			throw runtime_error(string("Expected \"//\" or \"// ---\" to terminate settings and source."));
 	}
+	// Register last source
```
Comment (cameel) [inline test/TestCaseReader.cpp]:
```suggestion
	// Register the last source as the main one
```

File: test/TestCaseReader.h [hunk 17]
```diff
@@ -32,7 +42,7 @@ class TestCaseReader
 	TestCaseReader() = default;
 	explicit TestCaseReader(std::string const& _filename);
 
-	std::map<std::string, std::string> const& sources() { return m_sources; }
+	SourceMap const& sources() { return m_sources; }
```
Comment (cameel) [inline test/TestCaseReader.h]:
```suggestion
	SourceMap const& sources() const { return m_sources; }
```

The functions below could be `const` too. None of them seems to modify any members and they return `const` values or copies.

File: test/libsolidity/SemanticTest.h [hunk 18]
```diff
@@ -58,9 +58,8 @@ class SemanticTest: public SolidityExecutionFramework, public EVMVersionRestrict
 	/// Compiles and deploys currently held source.
 	/// Returns true if deployment was successful, false otherwise.
 	bool deploy(std::string const& _contractName, u256 const& _value, bytes const& _arguments, std::map<std::string, solidity::test::Address> const& _libraries = {});
-
 private:
-	std::string m_source;
+	SourceMap m_source;
```
Comment (cameel) [inline test/libsolidity/SemanticTest.h]: For consistency with the reader class:
```suggestion
	SourceMap m_sources;
```

File: test/libsolidity/SemanticTest.cpp [hunk 19]
```diff
@@ -44,7 +44,7 @@ SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVer
 	EVMVersionRestrictedTestCase(_filename),
 	m_enforceViaYul(enforceViaYul)
 {
-	m_source = m_reader.source();
+	m_source = m_reader.sources();
 	m_lineOffset = m_reader.lineNumber();
```
Comment (cameel) [inline test/libsolidity/SemanticTest.cpp]: This could be on the constructor initializer list above.

File: test/libsolidity/SemanticTest.cpp [hunk 20]
```diff
@@ -241,7 +241,7 @@ TestCase::TestResult SemanticTest::run(ostream& _stream, string const& _linePref
 
 void SemanticTest::printSource(ostream& _stream, string const& _linePrefix, bool) const
 {
-	stringstream stream(m_source);
+	stringstream stream(m_source.sourceMap.at(m_source.mainSourceFile));
```
Comment (cameel) [inline test/libsolidity/SemanticTest.cpp]: Shouldn't this function print all the source files? Currently it prints only the main one.

File: test/libsolidity/SolidityExecutionFramework.cpp [hunk 21]
```diff
@@ -31,22 +31,46 @@ using namespace solidity::frontend;
...
 using namespace std;
 
+bytes SolidityExecutionFramework::multiSourceCompileContract(
+	map<string, string> const& _sourceCode,
+	string const& _contractName,
+	map<string, Address> const& _libraryAddresses
+)
+{
+	map<string, string> sourceCodeMap = _sourceCode;
+	for (auto &entry: sourceCodeMap)
```
Comment (cameel) [inline test/libsolidity/SolidityExecutionFramework.cpp]:
```suggestion
	for (auto& entry: sourceCodeMap)
```

File: test/libsolidity/SolidityExecutionFramework.cpp [hunk 22]
```diff
@@ -31,22 +31,46 @@ using namespace solidity::frontend;
...
-	// Silence compiler version warning
-	std::string sourceCode = "pragma solidity >=0.0;\n";
-	if (
-		solidity::test::CommonOptions::get().useABIEncoderV2 &&
-		_sourceCode.find("pragma experimental ABIEncoderV2;") == std::string::npos
-	)
-		sourceCode += "pragma experimental ABIEncoderV2;\n";
-	sourceCode += _sourceCode;
+	string sourceCode = _sourceCode;
+	addPreamble(sourceCode);
```
Comment (cameel) [inline test/libsolidity/SolidityExecutionFramework.cpp]: If you don't know that `addPreamble()` modifies its argument this makes you stop and think why a new variable is being introduced. I think a more explicit name would make the code much clearer:
```suggestion
	string sourceCodeWithPreamble = _sourceCode;
	addPreamble(sourceCodeWithPreamble);
```

File: test/libsolidity/SolidityExecutionFramework.cpp [hunk 23]
```diff
@@ -31,22 +31,46 @@ using namespace solidity::frontend;
...
+	return compileContractCommon(
+		{{ "", sourceCode }},
```
Comment (cameel) [inline test/libsolidity/SolidityExecutionFramework.cpp]:
```suggestion
		{{"", sourceCode}},
```

File: test/libsolidity/SolidityExecutionFramework.cpp [hunk 24]
```diff
@@ -85,3 +109,15 @@ bytes SolidityExecutionFramework::compileContract(
 		cout << "metadata: " << m_compiler.metadata(contractName) << endl;
 	return obj.bytecode;
 }
+
+void SolidityExecutionFramework::addPreamble(string& _sourceCode)
```
Comment (cameel) [inline test/libsolidity/SolidityExecutionFramework.cpp]: [Parameters used for input and output are supposed to have `io_` prefix (rather than `_`)](https://github.com/ethereum/solidity/blob/develop/CODING_STYLE.md#4-variable-prefixes):
```suggestion
void SolidityExecutionFramework::addPreamble(string& io_sourceCode)
```

Also, I think it would be better if this function returned a string instead of modifying its parameter. We can always use `move()` if we don't actually want to copy it.

File: test/libsolidity/SolidityExecutionFramework.cpp [hunk 25]
```diff
@@ -85,3 +109,15 @@ bytes SolidityExecutionFramework::compileContract(
...
+{
+	// Silence compiler version warning
+	string preamble = "pragma solidity >=0.0;\n";
+	if (
+		solidity::test::CommonOptions::get().useABIEncoderV2 &&
+		_sourceCode.find("pragma experimental ABIEncoderV2;") == string::npos
+	)
+		preamble += "pragma experimental ABIEncoderV2;\n";
+	_sourceCode = preamble + _sourceCode;
```
Comment (cameel) [inline test/libsolidity/SolidityExecutionFramework.cpp]: I think that it needs an explicit `move`:
```suggestion
	_sourceCode = preamble + move(_sourceCode);
```

File: test/libsolidity/SolidityExecutionFramework.cpp [hunk 26]
```diff
@@ -31,22 +31,46 @@ using namespace solidity::frontend;
...
+		_contractName,
+		_libraryAddresses
+	);
+}
+
+bytes SolidityExecutionFramework::compileContractCommon(
+	map<string, string> const& _sourceCodeMap,
+	string const& _contractName,
+	map<string, Address> const& _libraryAddresses
+)
```
Comment (cameel) [inline test/libsolidity/SolidityExecutionFramework.cpp]: This could be merged with `multiSourceCompileContract()`. `compileContract()` could call the multi-file version internally.

Maybe even it would be better to remove the single-file version as suggested in https://github.com/ethereum/solidity/pull/9228#discussion_r443707768.

File: test/libsolidity/SolidityExecutionFramework.h [hunk 27]
```diff
@@ -59,18 +59,44 @@ class SolidityExecutionFramework: public solidity::test::ExecutionFramework
...
+		u256 const& _value = 0,
+		std::string const& _contractName = "",
+		bytes const& _arguments = bytes(),
+		std::map<std::string, solidity::test::Address> const& _libraryAddresses = std::map<std::string, solidity::test::Address>()
+	) override
+	{
+		bytes bytecode = multiSourceCompileContract(_sources, _contractName, _libraryAddresses);
+		sendMessage(bytecode + _arguments, true, _value);
+		return m_output;
+	}
```
Comment (cameel) [inline test/libsolidity/SolidityExecutionFramework.h]: I think you should make `compileAndRunWithoutCheck()` call this function.

**Review by cameel — COMMENTED** [review]
Seems fine overall. I have lots of suggestions but mostly about improving the code.

The only thing that might be a real issue is `SemanticTest::printSource()` which I think should be printing multiple files now.

File: test/ExecutionFramework.h (hunk 2 above)
Comment (bshastry) [reply test/ExecutionFramework.h]: Ah I see, I changed it so that `compileAndRunWithoutCheck` defaults to a multi source version.

> Execution frameworks / compilers that do not support multi-source inputs can just assert that there is only one file.

Since both syntax and semantic tests now support multi-source inputs, I think this won't be necessary.

File: test/libsolidity/SolidityExecutionFramework.cpp (hunk 24 above)
Comment (bshastry) [reply test/libsolidity/SolidityExecutionFramework.cpp]: Switched to `move()` ing modified input as output

File: test/libsolidity/SemanticTest.cpp [hunk 28]
```diff
@@ -239,12 +237,51 @@ TestCase::TestResult SemanticTest::run(ostream& _stream, string const& _linePref
...
+	else
+		for (auto const& [name, source]: m_sources.sources)
+		{
+			if (outputSourceNames)
+				_stream << _linePrefix << "==== Source: " + name << " ====" << endl;
+			stringstream stream(source);
+			string line;
+			while (getline(stream, line))
+				_stream << _linePrefix << line << endl;
+		}
```
Comment (bshastry) [inline test/libsolidity/SemanticTest.cpp:282]: This is identical to https://github.com/ethereum/solidity/blob/e419e904d3ce2625995175dc155b911d2b4d2340/test/CommonSyntaxTest.cpp#L92

with the following lines (error parsing) removed

https://github.com/ethereum/solidity/blob/e419e904d3ce2625995175dc155b911d2b4d2340/test/CommonSyntaxTest.cpp#L110-L124

I didn't introduce error parsing with this PR because legacy code did not either. If we would like to support syntax error diagnosis for semantic tests, we should consider doing it here as well. But perhaps that is another issue.

**Comment by bshastry** [comment]
Thank you @cameel and @chriseth for your feedback. I have incorporated most of your comments. `printSource` for semantic tests now prints all sources (instead of only the main source file). However, it does not diagnose syntactic errors (if any) in semantic tests. This feature may be useful when one is initially writing a semantic test, however I feel it is a separate PR :)

**Comment by chriseth** [comment]
" it does not diagnose syntactic errors" - is that a regression?

File: test/libsolidity/SemanticTest.cpp [hunk 29]
```diff
@@ -41,12 +41,10 @@ namespace fs = boost::filesystem;
 
 SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVersion, bool enforceViaYul):
 	SolidityExecutionFramework(_evmVersion),
-	EVMVersionRestrictedTestCase(_filename),
+	EVMVersionRestrictedTestCase(_filename), m_sources(m_reader.sources()),
```
Comment (chriseth) [inline test/libsolidity/SemanticTest.cpp]:
```suggestion
	EVMVersionRestrictedTestCase(_filename),
	m_sources(m_reader.sources()),
```

File: test/libsolidity/SemanticTest.cpp [hunk 30]
```diff
@@ -239,12 +237,51 @@ TestCase::TestResult SemanticTest::run(ostream& _stream, string const& _linePref
...
-		_stream << _linePrefix << line << endl;
+	if (m_sources.sources.empty())
+		return;
+
+	bool outputSourceNames = true;
+	if (m_sources.sources.size() == 1 && m_sources.sources.begin()->first.empty())
+		outputSourceNames = false;
+
+	if (_formatted)
+		for (auto const& [name, source]: m_sources.sources)
```
Comment (chriseth) [inline test/libsolidity/SemanticTest.cpp]: Can you swap the `for` and the `if (_formatted)`?

File: test/libsolidity/SemanticTest.cpp (hunk 28 above)
Comment (chriseth) [reply test/libsolidity/SemanticTest.cpp:282]: I think it would make sense to use the same function for both.

At the same time, I would like to merge this rather sooner than later because we want to add some tests, so maybe it can be done in its own PR.

**Comment by bshastry** [comment]
> " it does not diagnose syntactic errors" - is that a regression?

I haven't checked git history but I doubt that it is a regression. I feel parsing errors in semantic tests were previously ignored.

Edit: I was mistaken. semantic test's `printSource` counterpart *does* diagnose syntactic errors. It only does not colour format the warnings/errors as syntactic test framework's `printSource` does.

I mistook the formatting code (orange text for warnings and red for errors) to be error diagnosis code. So please ignore my previous comment in this regard.

So what is missing in semantic test framework is syntactic error *colour highlighting*

File: test/TestCaseReader.cpp (hunk 15 above)
Comment (cameel) [reply test/TestCaseReader.cpp]: The code looks unchanged. Maybe you forgot to push this change?

**Comment by cameel** [comment]
They result in a failed assertion. Parsing errors are printed first though. Did that change now? It would definitely made debugging harder.

File: test/libsolidity/SolidityExecutionFramework.cpp [hunk 31]
```diff
@@ -85,3 +81,28 @@ bytes SolidityExecutionFramework::compileContract(
...
+{
+	return multiSourceCompileContract(
+		{{"", _sourceCode}},
+		_contractName,
+		_libraryAddresses
+	);
+}
+
+string SolidityExecutionFramework::addPreamble(string& _sourceCode)
```
Comment (cameel) [inline test/libsolidity/SolidityExecutionFramework.cpp]:
```suggestion
string SolidityExecutionFramework::addPreamble(string const& _sourceCode)
```
and then you'll also need to remove the `move()` below (it was only appropriate for the `string& io_sourceCode` version).

You could also keep the `move` and make the parameter `string _sourceCode` but that would only be better if you were modifying it in place. In my opinion `const&` is the best option here.

**Comment by bshastry** [comment]
> They result in a failed assertion. Parsing errors are printed first though. Did that change now? It would definitely made debugging harder.

Sorry for miscommunication on my part earlier, please see https://github.com/ethereum/solidity/pull/9228#issuecomment-648157803

File: test/libsolidity/SemanticTest.cpp (hunk 30 above)
Comment (bshastry) [reply test/libsolidity/SemanticTest.cpp]: I applied this change to the original code in `CommonSyntaxTest.cpp` as well

File: test/libsolidity/SolidityExecutionFramework.h [hunk 32]
```diff
@@ -47,30 +47,38 @@ class SolidityExecutionFramework: public solidity::test::ExecutionFramework
...
 
+	bytes multiSourceCompileContract(
+		std::map<std::string, std::string> const& _sources,
+		std::string const& _contractName = "",
+		std::map<std::string, solidity::test::Address> const& _libraryAddresses = {}
+	);
+
+	/// Returns @param _sourceCode prefixed with the version pragma and the ABIEncoderV2 pragma,
+	/// the latter only if it is required.
+	std::string addPreamble(std::string& _sourceCode);
```
Comment (cameel) [inline test/libsolidity/SolidityExecutionFramework.h]:
```suggestion
	static std::string addPreamble(std::string& _sourceCode);
```

File: test/TestCaseReader.cpp (hunk 15 above)
Comment (bshastry) [reply test/TestCaseReader.cpp]: I forgot to drop the comment. Will push an update.

File: test/libsolidity/SemanticTest.cpp (hunk 28 above)
Comment (bshastry) [reply test/libsolidity/SemanticTest.cpp:282]: Okay, let's move the refactoring to a future PR then.

Comment (bshastry) [reply test/libsolidity/SemanticTest.cpp:282]: May be that PR can also make error highlighting uniform across syntactic and semantic tests.

File: test/libsolidity/SemanticTest.cpp [hunk 33]
```diff
@@ -239,12 +238,50 @@ TestCase::TestResult SemanticTest::run(ostream& _stream, string const& _linePref
...
-	stringstream stream(m_source);
-	string line;
-	while (getline(stream, line))
-		_stream << _linePrefix << line << endl;
+	if (m_sources.sources.empty())
+		return;
+
+	bool outputSourceNames = true;
+	if (m_sources.sources.size() == 1 && m_sources.sources.begin()->first.empty())
+		outputSourceNames = false;
```
Comment (cameel) [inline test/libsolidity/SemanticTest.cpp]:
```suggestion
	bool outputSourceNames = (m_sources.sources.size() != 1 || !m_sources.sources.begin()->first.empty());
```

File: test/ExecutionFramework.h (hunk 11 above)
Comment (cameel) [reply test/ExecutionFramework.h]: How about the `{{"", _sourceCode}}` form? I see you used it in other places but here you're still using `std::map<std::string, std::string>{...}`

File: test/TestCaseReader.cpp (hunk 15 above)
Comment (cameel) [reply test/TestCaseReader.cpp]: Could you also change the `if` to `if (!currentSourceName.empty() || !currentSource.empty())`?

File: test/libsolidity/SolidityExecutionFramework.cpp (hunk 31 above)
Comment (bshastry) [reply test/libsolidity/SolidityExecutionFramework.cpp]: sounds good.

File: test/libsolidity/SolidityExecutionFramework.cpp (hunk 22 above)
Comment (cameel) [reply test/libsolidity/SolidityExecutionFramework.cpp]: This is no longer relevant after other changes so I'm marking it as resolved.

File: test/TestCaseReader.h (hunk 17 above)
Comment (cameel) [reply test/TestCaseReader.h]: What about the functions below? Could you change them too? Or was there just something preventing that change?

File: test/libsolidity/SemanticTest.cpp (hunk 33 above)
Comment (bshastry) [reply test/libsolidity/SemanticTest.cpp]: Ported change to `CommonSyntaxTests.cpp` as well.

File: test/libsolidity/SemanticTest.cpp (hunk 28 above)
Comment (bshastry) [reply test/libsolidity/SemanticTest.cpp:282]: created https://github.com/ethereum/solidity/issues/9250 to track this

File: test/libsolidity/SolidityExecutionFramework.cpp [hunk 34]
```diff
@@ -31,22 +31,18 @@ using namespace solidity::frontend;
...
-	// Silence compiler version warning
-	std::string sourceCode = "pragma solidity >=0.0;\n";
-	if (
-		solidity::test::CommonOptions::get().useABIEncoderV2 &&
-		_sourceCode.find("pragma experimental ABIEncoderV2;") == std::string::npos
-	)
-		sourceCode += "pragma experimental ABIEncoderV2;\n";
-	sourceCode += _sourceCode;
+	map<string, string> sources = _sourceCode;
+	for (auto& entry: sources)
```
Comment (cameel) [inline test/libsolidity/SolidityExecutionFramework.cpp]:
```suggestion
	map<string, string> sourcesWithPreamble = _sourceCode;
	for (auto& entry: sourcesWithPreamble)
```

**Review by cameel — DISMISSED** [review]
I have one more suggestion and I think you have missed https://github.com/ethereum/solidity/pull/9228#discussion_r443763723 (github folds it under `Load more...`) but I'm already marking as approved because all the more important points have been addressed and it looks fine overall.

**Review by cameel — DISMISSED** [review]

File: test/TestCaseReader.h (hunk 17 above)
Comment (bshastry) [reply test/TestCaseReader.h]: No, there was nothing preventing the change. I may have overlooked your second comment though.

**Review by chriseth — DISMISSED** [review]

**Comment by bshastry** [comment]
> I have one more suggestion and I think you have missed [#9228 (comment)](https://github.com/ethereum/solidity/pull/9228#discussion_r443763723) (github folds it under `Load more...`) but I'm already marking as approved because all the more important points have been addressed and it looks fine overall.

Done :) Thank you for the reviews!

**Review by cameel — DISMISSED** [review]

**Review by chriseth — APPROVED** [review]
